data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,203 @@
+
+    print(f"Table {table_name} successfully written to BigQuery")
+
+
+def print_df(df, table_name):
+    '''Print 20 rows from dataframe and a random sample'''
+    # first 100 rows for smaller tables
+    df.show()
+
+    # random sample for larger tables
+    # for small tables this will be empty
+    df.sample(True, 0.0001).show(n=500, truncate=False)
+
+    print(f"Table {table_name} printed")
+
+
+def main():
+    # Create a SparkSession under the name "setup". Viewable via the Spark UI
+    spark = SparkSession.builder.appName("setup").getOrCreate()
+
+    upload = True  # Whether to upload data to BigQuery
+
+    # Check whether or not results should be uploaded
+    if '--test' in sys.argv:
+        upload = False
+        print("Not uploading results to BigQuery")
+    else:
+        create_bigquery_dataset()
+        print("Results will be uploaded to BigQuery")
+
+    # Ingest External Datasets
+
+    for table_name, data in EXTERNAL_DATASETS.items():
+        print(f'Creating dataframe for {table_name}')
+        df = spark.createDataFrame(pd.read_csv(data["url"]),
+                                   schema=data["schema"])
+
+        if upload:
+            write_to_bigquery(df, table_name)
+        else:
+            print_df(df, table_name)
+
+    # Check if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+    except Py4JJavaError:
+        print(f"{TABLE} does not exist. ")
+        return
+
+    # Declare dictionary with keys column names and values user defined
+    #  functions and return types
+    udf_map = {
+            'tripduration': (trip_duration, StringType()),
+            'start_station_name': (station_name, StringType()),
+            'start_station_latitude': (convert_angle, StringType()),
+            'start_station_longitude': (convert_angle, StringType()),
+            'end_station_name': (station_name, StringType()),
+            'end_station_latitude': (convert_angle, StringType()),
+            'end_station_longitude': (convert_angle, StringType()),
+            'usertype': (user_type, StringType()),
+            'gender': (gender, StringType()),
+    }
+
+    # Declare which columns to set some values to null randomly
+    null_columns = [
+            'tripduration',
+            'starttime',
+            'stoptime',
+            'start_station_latitude',
+            'start_station_longitude',
+            'end_station_latitude',
+            'end_station_longitude',
+    ]
+
+    # Dirty the columns
+    for name, udf in udf_map.items():
+        df = df.withColumn(name, UserDefinedFunction(*udf)(name))
+
+    # Randomly set about 5% of the values in some columns to null
+    for name in null_columns:
+
+        df = df.withColumn(name, when(expr("rand() < 0.05"), None).otherwise(df[name]))
+
+    # Duplicate about 0.01% of the rows
+    dup_df = df.sample(True, 0.0001)
+
+    # Create final dirty dataframe
+    df = df.union(dup_df)
+
+    if upload:
```

Comment:
What's the difference between this upload and write to bigquery and the other? 

You may want to consider also making this a function because it's repeated code.